File: test/matlab/matrix_rw__test.m
%* setup
bin_fprefix = '../../build/bin';
test_name = 'matrix_rw-test';
bin_fname = append(test_name, '.exe');
fprefix = append('../../build/data/', test_name, '-');
ref_fprefix = append('../ref_data/', test_name, '-');
mat_fname = 'mat.csv';
ref_mat_fname = 'ref_mat.csv';
m_col = 75;
is_single_precision = false;

if is_single_precision
	error_thres = 1e-7; %* single precision
else 
	error_thres = 1e-15;
end

%* create reference data 
magic_mat = magic(m_col);
ref_mat = [magic_mat; magic_mat]./max(magic_mat, [], 'all') - 1/3;
writematrix(ref_mat, append(ref_fprefix, ref_mat_fname), 'Delimiter', ',');
disp(append('Created reference data for ', test_name));

prev_pwd = pwd;
cd(bin_fprefix);

if isfile(bin_fname)
	%* call the test executable
	if system(bin_fname) > 0
		warning(append(bin_fprefix, '/', bin_fname, ' has returned failure.'));
	end
	%* read the results
	mat = readmatrix(append(fprefix, mat_fname));
	%* verify
	max_error = max(mat - ref_mat, [], 'all')

	if max_error < error_thres
		disp(append(test_name, '	ok'));
	else
		disp(append(test_name, '	fail'));
	end
else
	error(append(bin_fprefix, '/', bin_fname, ' does not exist. Use CMake to build the test.'));
end
cd(prev_pwd);

File: test/matrix_rw-test.cpp
#include "matrix_rw.hpp"
#include <limits> /** std::numeric_limits */
#include <string>

using matrix_rw::Real_T;
using matrix_rw::size_t;
using matrix_rw::VarRowMat_T;

constexpr size_t m_col = 75;
constexpr size_t n_row_default = 150;
constexpr Real_T error_threshold = std::numeric_limits<Real_T>::epsilon();

//* relative to the directory of the executable
const std::string test_name = "matrix_rw-test";
const std::string fpath = "../data/" + test_name + "-";
const std::string ref_fpath = "../../test/ref_data/" + test_name + "-";
const std::string mat_ref_fname = "ref_mat.csv";
const std::string mat_fname = "mat.csv";

int
main()
{
	matrix_rw::Reader<m_col> read;
	matrix_rw::Writer<m_col> write;

	/** read the reference data */
	VarRowMat_T<m_col> mat_ref;
	mat_ref.reserve(n_row_default);
	read(ref_fpath + mat_ref_fname, mat_ref); /** read from file */
	const size_t n_row = mat_ref.size();

	/** write the data back */
	write(fpath + mat_fname, mat_ref);

	/** read it again */
	VarRowMat_T<m_col> mat;
	mat.reserve(n_row);
	read(fpath + mat_fname, mat);

	/** verify */
	Real_T max_error = 0;

	for (size_t i = 0; i < n_row; ++i) {
		for (size_t j = 0; j < m_col; ++j) {
			const Real_T error = std::abs(mat[i][j] - mat_ref[i][j]);

			if (error > max_error) {
				max_error = error;
			}
		}
	}

	if (n_row > 0 && max_error < error_threshold) {
		return 0;
	} else {
		return 1;
	}
}